### DISCUSSION TITLE
Unable to clear Listbox selection

### DISCUSSION THREAD

**Discussion by twisted89** [AvaloniaUI/Avalonia#8546] [Q&A]
I can't seem to reset the selected item on a Listbox. 
Setting the selected item to null or selected index to -1 does nothing, even directly calling Selection.Clear() on the listbox does nothing.  

I'm using the following code

```
<ListBox x:Name="AccountList" Items="{Binding Accounts}" ScrollViewer.HorizontalScrollBarVisibility="Disabled" SelectedItem="{Binding SelectedAccount, Mode=TwoWay}" >
  <ListBox.DataTemplates>
  <views:AccountSelector>
	  <DataTemplate x:Key="AccountType1">
		  <styles:Card MinWidth="240" MaxHeight="80" MinHeight="80" VerticalAlignment="Top" Margin="0">
			  <StackPanel Orientation="Horizontal">
				  <Svg Path="/Assets/account1.svg" Width="50" Height="50"/>
				  <StackPanel Margin="16,0" >
					  <TextBlock Margin="0,12,0,0" TextWrapping="Wrap" Text="{ReflectionBinding Name}"/>
					  <TextBlock Margin="0,12,0,0" TextWrapping="Wrap" Text="{ReflectionBinding TypeName}"/>
				  </StackPanel>
			  </StackPanel>
		  </styles:Card>
	  </DataTemplate>
	  <DataTemplate x:Key="AccountType2">
		  <styles:Card MinWidth="240" MaxHeight="80" MinHeight="80" VerticalAlignment="Top" Margin="0">
			  <StackPanel Orientation="Horizontal">
				  <Svg Path="/Assets/account2.svg" Width="50" Height="50"/>
				  <StackPanel Margin="16,0" Orientation="Vertical">
					  <TextBlock Margin="0,12,0,0" TextWrapping="Wrap" Text="{ReflectionBinding Name}"/>
					  <TextBlock Margin="0,12,0,0" TextWrapping="Wrap" Text="{ReflectionBinding TypeName}"/>
				  </StackPanel>
			  </StackPanel>
		  </styles:Card>
	  </DataTemplate>
  </views:AccountSelector>
  </ListBox.DataTemplates>
  <ListBox.ItemsPanel>
  <ItemsPanelTemplate>
	  <WrapPanel Orientation="Horizontal"  />
  </ItemsPanelTemplate>
  </ListBox.ItemsPanel>
  </ListBox>
```

The last clicked item will always be highlighted and clicking it again doesn't trigger selectionchanged even though the selected item was reset to null. 

The only way I've been able to clear the selection is completely resetting the accounts binding list. 

Any help would be appreciated.

**Comment by timunie** [comment]
I'm wondering what is going on here. I can't see this behavior in my App. 

Maybe you can play around with different SelectionModes

**Reply by twisted89** [reply]
So if you clear the selected item on the listbox you can click the same item again and it fires the selecteditemchanged event? 
The closest I can get is setting the selection mode to Toggle, but to fire the selection changed when clicking the same item again I have to click twice. First time deselects it and changes selecteditem to null, second time it selects the item again.

I traced back through the Avalonia code and I can't see how it's ever meant to work. 

In the OnPointerPressed function for Avalonia.Controls.ListBox it always calls "UpdateSelectionFromEventSource" with the select parameter set to true.

I can't see anywhere to deselect the current selection without the mode being set the toggle and clicking twice....

**Reply by timunie** [reply]
Selecting a selected item again doesn't fire selection changed as there was no change in the selection. What exactly do you want to achieve?

**Reply by twisted89** [reply]
As I said in the OP "I can't seem to reset the selected item on a Listbox".
The issue here is that selecteditem is set to null yet clicking the same item doesn't do anything.

I need to clear the selection so selecteditem is null, then if a user clicks the same item again, re-fire the selection changed event. Or, at a bare minimum know that the user has re-selected the same item which I can't see any way to do either.

**Reply by timunie** [reply]
Can you upload a minimal sample to GitHub? I wonder if some Binding does not work well. I never had such a problem on my own, but probably I oversee something.

**Reply by timunie** [reply]
I added some debug output and it shows: 

```cs
private void ListBox_OnSelectionChanged(object? sender, SelectionChangedEventArgs e)
{
   Debug.WriteLine($"Selection changed; {e.AddedItems.Count} items added, {e.RemovedItems.Count} removed.");
}
```

Output:
```
Selection changed; 0 items added, 1 removed.
Selection changed; 1 items added, 0 removed.
Selection changed; 1 items added, 1 removed.
Selection changed; 1 items added, 1 removed.
Selection changed; 0 items added, 1 removed.
Selection changed; 1 items added, 0 removed.
```

**Reply by twisted89** [reply]
Example project attached.
[SelectTest.zip](https://github.com/AvaloniaUI/Avalonia/files/9150069/SelectTest.zip)

**Reply by timunie** [reply]
I think you should handle the logic inside your data template. I've done a few tweaks and it works for me without any issue. Also I removed the data template selector as you don't need it in such a simple scenario.

Attached please find my proposal. 

[SelectTest.zip](https://github.com/AvaloniaUI/Avalonia/files/9166674/SelectTest.zip)

**Reply by twisted89** [reply]
I've managed to identify the exact cause of the issue - If you subscribe to the viewmodel property changed event and set the selected account to null in there it prevents you from reselecting the same account. Your example worked because you're clearing the selection with a button that happens outside the property changed event. 

```
ViewModel.PropertyChanged += Property_Changed;

private void Property_Changed(object sender, PropertyChangedEventArgs e)
{
    if (e.PropertyName == "SelectedAccount" && ViewModel.SelectedAccount != null)
    {
        ViewModel.SelectedAccount = null;
    }
}
```

 More of a workaround than a fix but posting to the ui thread avoids the issue

`Dispatcher.UIThread.Post(() => { ViewModel.ClearCommand(); });`
